libs/constrained-generators/src/Constrained/Spec/Map.hs

```diff
@@ -308,7 +350,537 @@ dom_ ::
+      , (1, pure Nothing)
+      , (1, Just . MinValSize . MemberSpec . (mustValLen :) <$> vectorOf 5 genOk)
+      ]
+  mustval <- vectorOf (fromInteger mustValLen) arbitrary
+  let avg =
+        if null mustval
+          then 0 :: Integer
+          else sum mustval `div` toInteger (length mustval)
+      goodMust = constrained (\n -> n >=. (lit avg))
+  mustkey <- genFromGenT $ (Set.fromList <$> genDistinctList @_ @fn goodMust mustKeyLen [])
+  let kvSpec =
+        constrained
+          ( \p ->
+              match p (\key val -> [assert $ key >=. val])
+          )
+  pure (MapSpec Nothing mustkey mustval mapsize kvSpec NoFold valsize)
+
+maxSize :: Specification fn Integer -> Integer
+maxSize (TypeSpec (NumSpecInterval _ (Just n)) _) = n
+maxSize (TypeSpec (NumSpecInterval _ Nothing) _) = 8
+maxSize (MemberSpec xs) = maximum xs
+maxSize _ = 9
+
+-- | Gen 3 Integers They correspond to the
+--   1) size of the mustKeys,
+--   2) size of the mustVals
+--   3) The minimum size, one might get from the SizeSpec for the Map
+--   All of them could be 0
+genSizes :: Gen (Integer, Integer, Integer)
+genSizes =
+  genFromGenT
+    ( genFromSpecT @BaseFn
+        ( constrained
+            ( \triple ->
+                match
+                  triple
+                  ( \k v slack ->
+                      [ assert $ 0 <=. v
+                      , assert $ v <=. 6
+                      , assert $ 0 <=. k
+                      , assert $ k <=. 7
+                      , assert $ 0 <=. slack
+                      , assert $ slack <=. 5
+                      ]
+                  )
+            )
+        )
+    )
+
+-- ==============================================================================
+
+-- Like conformsTo, but tells which of the sub conformance components fail
+-- Nothing means success, (Just s), means 's' tells what went wrong.
+conformsToMapMaybe ::
+  forall fn k v.
+  (HasSpec fn (Map k v), HasSpec fn k, HasSpec fn v, Ord k) =>
+  Map k v ->
+  MapSpec fn k v ->
+  Maybe String
+conformsToMapMaybe m (MapSpec _ mustKeys mustVals size kvs foldSpec minval) =
+  allOf
+    [ tell "mustKeys" $ mustKeys `Set.isSubsetOf` Map.keysSet m
+    , tell "mustVals" $ all (`elem` Map.elems m) mustVals
+    , tell ("size (" ++ show size ++ ") on map with size" ++ show (Map.size m)) $
+        sizeOf m `conformsToSpec` size
+    , tell "kvs" $ all (`conformsToSpec` kvs) (Map.toList m)
+    , tell "foldSpec" $ Map.elems m `conformsToFoldSpec` foldSpec
+    , tell "minValSize" $ conformsToSpec @fn (sizeOf (nub (Map.elems m))) (extractFromMinValSize minval)
+    ]
+  where
+    tell :: String -> Bool -> Maybe String
+    tell _ True = Nothing
+    tell msg False = Just msg
+
+    allOf :: [Maybe String] -> Maybe String
+    allOf xs = foldr accum Nothing xs
+      where
+        accum Nothing ans = ans
+        accum (Just s) Nothing = Just s
+        accum (Just s) (Just more) = Just (s ++ "\n" ++ more)
```

Comment:
Perhaps better to do this using the `GE` machinery we already use extensively in this code base?

Reply:
That is not what I am getting at.  When come tries to compute the most accurate Size for a MapSpec,
many things constribute to the answer. If one of those things is incossitent, there is no way to tell. Things just return ErrorSpec. This  labels aech thing and reports which ones fail.